lib/tasks_private/instance_types.rake

```diff
@@ -1,215 +1,94 @@
-
-  new_instance_types_file << output_instances.map { |i| format_instance(i) }.join("\n")
-  new_instance_types_file << <<EOT
-  }.freeze
-
-  # Types that are no longer advertised
-  DISCONTINUED_TYPES = {
-EOT
-  # current discontinued types
-  output_instances = []
-  current_discontinued = ManageIQ::Providers::Amazon::InstanceTypes::DISCONTINUED_TYPES.dup
-  current_discontinued.values.each do |instance|
-    instance[:disabled] = true
-    output_instances << instance
-  end
+    old_types_data = YAML.load_file(out_file)
+    default_type = old_types_data.find { |_, data| data[:default] }.first
+    discontinued_types = (old_types_data.keys - types_data.keys)
+    types_data.merge!(old_types_data.slice(*discontinued_types))
 
-  # new discontinued from current_available
-  discontinued_count = current_available.count
-  current_available.values.each do |instance|
-    instance[:disabled] = true
-    output_instances << instance
-  end
+    ## postprocess
 
-  # new discontinued from current_deprecated
-  discontinued_count += current_deprecated.count
-  current_deprecated.values.each do |instance|
-    instance[:disabled] = true
-    output_instances << instance
-  end
+    types_data.each do |instance_type, instance_data|
+      instance_data.except!(:default, :deprecated, :discontinued, :disabled)
+      instance_data[:default] = true if instance_type == default_type
+      if discontinued_types.include?(instance_type) || !instance_data[:current_version]
+        instance_data[:discontinued] = true
+        instance_data[:disabled] = true
+      end
+      if instance_data[:current_version] && !instance_data[:current_generation]
+        instance_data[:deprecated] = true
+      end
+    end
 
-  new_instance_types_file << output_instances.map { |i| format_instance(i) }.join("\n")
-  new_instance_types_file << <<EOT
-  }.freeze
+    types_data.sort!
+    types_data.each_value(&:sort!)
 
-  def self.instance_types
-    additional = Hash(Settings.ems.ems_amazon.try!(:additional_instance_types)).stringify_keys
-    disabled = Array(Settings.ems.ems_amazon.try!(:disabled_instance_types))
-    AVAILABLE_TYPES.merge(DEPRECATED_TYPES).merge(DISCONTINUED_TYPES).merge(additional).except(*disabled)
-  end
+    ## show warnings
 
-  def self.all
-    instance_types.values
-  end
+    unless collecting_warnings.empty?
+      info = collecting_warnings.transform_keys(&:first)
+      info.each_value do |instance_data|
+        instance_data.transform_values!(&:to_a)
+      end
+      info.sort!
+      log_data('Attention! Contradictory products data:', info)
+    end
+    log_data('Attention! Unforeseen values format:', parsing_warnings)
 
-  def self.names
-    instance_types.keys
-  end
-end
-EOT
+    ## save data
 
-  File.open(ManageIQ::Providers::Amazon::Engine.root.join('instance_types.rb'), 'w') do |f|
-    f.puts(new_instance_types_file)
+    out_file.write(types_data.to_yaml.each_line.map(&:rstrip).join("\n") << "\n")
```

Comment:
I would avoid the line stripping...let `to_yaml` do what it does and just write that.  As long as there is consistency, autogenerated files, IMO, should be as close to raw as possible.

So this line simplifies to `outfile.write(types_data.to_yaml)`

Reply:
rubocop complains on right spaces

Reply:
autogenerated files should be ignored by rubocop...we just put it in the exclusion list